src/vanilla.ts

```diff
@@ -45,44 +25,50 @@ export type StoreApi<T extends State> = {
   subscribe: Subscribe<T>
   destroy: Destroy
 }
+
 export type StateCreator<
   T extends State,
-  CustomSetState = SetState<T>,
-  CustomGetState = GetState<T>,
-  CustomStoreApi extends StoreApi<T> = StoreApi<T>
-> = (set: CustomSetState, get: CustomGetState, api: CustomStoreApi) => T
-
-function createStore<
-  TState extends State,
-  CustomSetState,
-  CustomGetState,
-  CustomStoreApi extends StoreApi<TState>
->(
-  createState: StateCreator<
-    TState,
-    CustomSetState,
-    CustomGetState,
-    CustomStoreApi
-  >
-): CustomStoreApi
-
-function createStore<TState extends State>(
-  createState: StateCreator<TState, SetState<TState>, GetState<TState>, any>
-): StoreApi<TState>
-
-function createStore<
-  TState extends State,
-  CustomSetState,
-  CustomGetState,
-  CustomStoreApi extends StoreApi<TState>
+  Mis extends [StoreMutatorIdentifier, unknown][],
+  Mos extends [StoreMutatorIdentifier, unknown][],
```

Comment:
Can these have defaults to avoid breaking change?
If it's not possible, we need to make sure to warn it in the release note.

```suggestion
  Mis extends [StoreMutatorIdentifier, unknown][] = [],
  Mos extends [StoreMutatorIdentifier, unknown][] = [],
```

Reply:
Hmm, I try to avoid defaults but I think we can have them here for backward compatibility. Although some places where `StateCreator` is used, for eg middlewares, there will be migration needed.

Reply:
I don't like defaults either, if possible.
What about we define two types?
```ts
export type StateCreator<
  T extends State,
> = ...
export type StateCreatorWithMutator<
  T extends State,
  Mis extends [StoreMutatorIdentifier, unknown][],
  Mos extends [StoreMutatorIdentifier, unknown][],
  U = T
> = ...
```

Reply:
No I think two types are much worse than defaults because then we'd have to educate users when to use which. Moreover we'll have two types that mostly do the same thing. I'd prefer defaults here.

> Maybe the same thing?

I don't follow what you mean by that haha.

If you're saying that if migration is needed even with defaults then the defaults are not worth much? Well yeah kinda. Do you have cases where users would be using `StateCreator` apart from middlewares? So I can understand how much migration we will save when we introduce defaults.

Reply:
You understand what I mean correctly.

Yeah, I'm not quite sure either how many users depend on `StateCreator` type.
It's me that introduced the `Custom` things and it doesn't work well as I originally hoped.
And, this PR fixes it.

So, I now feel like this can be a new type, like you did so in #662.
How about deprecating `StateCreator` in v3 (#715)?
```ts
/**
 * @deprecated `StateCreator` is deprecated and will be removed in v4.
 * If necessary, please define it on your end.
 * type StateCreator<T extends State> = (set: SetState<T>, get: GetState<T>, api: StoreApi<T>) => T
 * /
```

And, in v4 (#725), we use a new name. I would say `CreateState`, but your original suggestion `StoreInitializer` is possible.

Reply:
I don't think that solves the problem. For example users might be using `StateCreator` while [creating slices that the wiki suggests](https://github.com/pmndrs/zustand/wiki/Splitting-the-store-into-separate-slices). This is a legitimate usecase of `StateCreator`, if we deprecate it they'll define their own `StateCreator` in meantime and then again in v4 they'll have to use the new `CreateState`/`StoreIntializer` type. All this is a little messy.

I think in this case it's okay to have the `= []` defaults, they seem reasonable. It won't remove the necessity of migration, there's no way to escape breaking changes, the best we can do is making them minimal which the defaults do.

Reply:
Okay, then. Let's go with `= []`.